In [1]:
from google.colab import userdata

In [2]:
READ_TOKEN = userdata.get('HF_READ')

In [3]:
WRITE_TOKEN = userdata.get('HF_WRITE')

This command

In [4]:
!git config --global credential.helper store

In [5]:
!huggingface-cli login

⚠️  Warning: 'huggingface-cli login' is deprecated. Use 'hf auth login' instead.

    _|    _|  _|    _|    _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|_|_|_|    _|_|      _|_|_|  _|_|_|_|
    _|    _|  _|    _|  _|        _|          _|    _|_|    _|  _|            _|        _|    _|  _|        _|
    _|_|_|_|  _|    _|  _|  _|_|  _|  _|_|    _|    _|  _|  _|  _|  _|_|      _|_|_|    _|_|_|_|  _|        _|_|_|
    _|    _|  _|    _|  _|    _|  _|    _|    _|    _|    _|_|  _|    _|      _|        _|    _|  _|        _|
    _|    _|    _|_|      _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|        _|    _|    _|_|_|  _|_|_|_|

    A token is already saved on your machine. Run `hf auth whoami` to get more information or `hf auth logout` if you want to log out.
    Setting a new token will erase the existing one.
    To log in, `huggingface_hub` requires a token generated from https://huggingface.co/settings/tokens .
Enter your token (input will not be visible): 
Add t

In [6]:
!huggingface-cli whoami

⚠️  Warning: 'huggingface-cli whoami' is deprecated. Use 'hf auth whoami' instead.
AtharvaBOT7


In [7]:
!huggingface-cli repo create hftesting --repo-type model

⚠️  Warning: 'huggingface-cli repo' is deprecated. Use 'hf repo' instead.
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_http.py", line 409, in hf_raise_for_status
    response.raise_for_status()
  File "/usr/local/lib/python3.12/dist-packages/requests/models.py", line 1026, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 409 Client Error: Conflict for url: https://huggingface.co/api/repos/create

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/huggingface-cli", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/huggingface_hub/commands/huggingface_cli.py", line 61, in main
    service.run()
  File "/usr/local/lib/python3.12/dist-packages/huggingface_hub/commands/repo.py", line 141, in run
    repo_url = self._api.create_repo(
           

In [8]:
import os
from huggingface_hub import HfApi
from google.colab import userdata

In [9]:
api = HfApi(token=WRITE_TOKEN)

In [10]:
# We can create a repo using API option as well using this command

# api.create_repo(
#     repo_id="AtharvaBOT7/testinghf",
#     repo_type="model",
#     private=False
# )

In [11]:
# We can upload a file using this command using API

# api.upload_file(
#     path_or_fileobj="./test.txt",
#     path_in_repo="test.txt",
#     repo_id="AtharvaBOT7/testinghf",
#     repo_type="model"
# )

In [12]:
# We can also perform a notebook login on HuggingFace

from huggingface_hub import notebook_login

notebook_login()

### Installing Datasets

fsspec - file and system specifications

datasets - to install the datasets from huggingface.

In [13]:
!pip install -U fsspec "datasets<3.0.0"

  Using cached fsspec-2025.9.0-py3-none-any.whl.metadata (10 kB)


In [14]:
# Import the load_dataset function from datasets library
from datasets import load_dataset

In [15]:
dataset = load_dataset("stanfordnlp/imdb")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [16]:
print(dataset)

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 25000
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 25000
    })
    unsupervised: Dataset({
        features: ['text', 'label'],
        num_rows: 50000
    })
})


In [17]:
print(dataset['train'][0])

{'text': 'I rented I AM CURIOUS-YELLOW from my video store because of all the controversy that surrounded it when it was first released in 1967. I also heard that at first it was seized by U.S. customs if it ever tried to enter this country, therefore being a fan of films considered "controversial" I really had to see this for myself.<br /><br />The plot is centered around a young Swedish drama student named Lena who wants to learn everything she can about life. In particular she wants to focus her attentions to making some sort of documentary on what the average Swede thought about certain political issues such as the Vietnam War and race issues in the United States. In between asking politicians and ordinary denizens of Stockholm about their opinions on politics, she has sex with her drama teacher, classmates, and married men.<br /><br />What kills me about I AM CURIOUS-YELLOW is that 40 years ago, this was considered pornographic. Really, the sex and nudity scenes are few and far be

In [18]:
print(dataset['train'].features)
print(dataset['train'].column_names)
print(dataset['train'].features['label'].names)
print(dataset['train'].num_rows)
print(dataset['train'].num_columns)
print(dataset['train'].shape)

{'text': Value(dtype='string', id=None), 'label': ClassLabel(names=['neg', 'pos'], id=None)}
['text', 'label']
['neg', 'pos']
25000
2
(25000, 2)


We use the seed parameter inside the shuffle function to keep the shuffling same, it is like giving an id to that shuffle and it will create the same shuffle in each and every machine.

In [19]:
shuffled = dataset['train'].shuffle(seed=42)

The select range function will select random 100 values from the dataset.

In [20]:
shuffled.select(range(100))

Dataset({
    features: ['text', 'label'],
    num_rows: 100
})

Taking out 5000 rows as a training dataset.

In [21]:
subset_train = dataset['train'].shuffle(seed=42).select(range(5000))

In [22]:
subset_train

Dataset({
    features: ['text', 'label'],
    num_rows: 5000
})

## Performing Data Preprocessing

In [23]:
dataset['train']

Dataset({
    features: ['text', 'label'],
    num_rows: 25000
})

In [24]:
# Filtering the dataset to keep only the short reviews

short_reviews = dataset['train'].filter(lambda x: len(x['text']) <100)

In [25]:
short_reviews

Dataset({
    features: ['text', 'label'],
    num_rows: 9
})

In [26]:
# Filtering the dataset to keep only the short reviews

short_reviews = subset_train.filter(lambda x: len(x['text']) <100)

In [27]:
short_reviews

Dataset({
    features: ['text', 'label'],
    num_rows: 2
})

In [28]:
# Taking out the short positive reviews

short_positive = dataset['train'].filter(lambda x: x['label'] == 1 and len(x['text']) < 200)

In [29]:
short_positive

Dataset({
    features: ['text', 'label'],
    num_rows: 60
})

We are creating a new function to count the number of tokens in that particular review.

Then we will add a new column to our dataset with this parameter of length for better pre-processing.

In [30]:
def add_word_count(example):
  example["word_count"] = len(example["text"].split())
  return example

In [31]:
dataset = dataset.map(add_word_count)

In [32]:
dataset

DatasetDict({
    train: Dataset({
        features: ['text', 'label', 'word_count'],
        num_rows: 25000
    })
    test: Dataset({
        features: ['text', 'label', 'word_count'],
        num_rows: 25000
    })
    unsupervised: Dataset({
        features: ['text', 'label', 'word_count'],
        num_rows: 50000
    })
})

In [33]:
print(dataset['train'][1]['word_count'])

214


Now we will learn about streaming huggingface. When the size of the dataset is large enough and we do not have space on our disk, we can stream the dataset sample by sample on the fly.

In [34]:
big_data = load_dataset("openwebtext", streaming=True)

In [35]:
big_data

IterableDatasetDict({
    train: IterableDataset({
        features: ['text'],
        n_shards: 21
    })
})

IterableDataset - Dictionary like object with splits

train - only split provided in openwebtext

features: ["text"] - each example is a dictionary

n_shards - internally the dataset is chunked into 21 streamable parts

In [36]:
for item in big_data['train'].take(5):
  print(item)

{'text': 'Port-au-Prince, Haiti (CNN) -- Earthquake victims, writhing in pain and grasping at life, watched doctors and nurses walk away from a field hospital Friday night after a Belgian medical team evacuated the area, saying it was concerned about security.\n\nThe decision left CNN Chief Medical Correspondent Sanjay Gupta as the only doctor at the hospital to get the patients through the night.\n\nCNN initially reported, based on conversations with some of the doctors, that the United Nations ordered the Belgian First Aid and Support Team to evacuate. However, Belgian Chief Coordinator Geert Gijs, a doctor who was at the hospital with 60 Belgian medical personnel, said it was his decision to pull the team out for the night. Gijs said he requested U.N. security personnel to staff the hospital overnight, but was told that peacekeepers would only be able to evacuate the team.\n\nHe said it was a "tough decision" but that he accepted the U.N. offer to evacuate after a Canadian medical t

In [37]:
# This will store the dataset's english version and it will only choose the training subset of the data.

# web_crawled_data = load_dataset("c4", "en", split="train", streaming=True)

In [38]:
# this will download the dataset "tweet_eval" and it will only load the "sentiment" label from the dataset

dataset = load_dataset("tweet_eval", "sentiment")

In [39]:
dataset

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 45615
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 12284
    })
    validation: Dataset({
        features: ['text', 'label'],
        num_rows: 2000
    })
})

In [40]:
from collections import Counter

label_counts = Counter([ex["label"] for ex in dataset["train"]])
labels = ["Negative", "Neutral", "Positive"]
counts = [label_counts[i] for i in range(3)]

In [41]:
counts

[7093, 20673, 17849]

## Creating a Custom Dataset and Pushing it to HuggingFace

In [42]:
from datasets import Dataset
import pandas as pd

# Randomly generated values
data = {
    "text": [
      "The quick brown fox jumps over the lazy dog.",
      "Artificial intelligence is transforming the world.",
      "Hugging Face makes working with NLP easier.",
      "Streaming datasets allow training on large corpora.",
      "Python is my favorite programming language.",
      "Open source contributions accelerate research.",
      "Data engineers love clean and structured pipelines.",
      "Transformers are powerful models for many tasks.",
      "Machine learning requires both theory and practice.",
      "Randomly generated text can still be useful for testing."
    ],
    "label": [0, 1, 2, 0, 1, 2, 0, 1, 2, 0]
}

In [43]:
data

{'text': ['The quick brown fox jumps over the lazy dog.',
  'Artificial intelligence is transforming the world.',
  'Hugging Face makes working with NLP easier.',
  'Streaming datasets allow training on large corpora.',
  'Python is my favorite programming language.',
  'Open source contributions accelerate research.',
  'Data engineers love clean and structured pipelines.',
  'Transformers are powerful models for many tasks.',
  'Machine learning requires both theory and practice.',
  'Randomly generated text can still be useful for testing.'],
 'label': [0, 1, 2, 0, 1, 2, 0, 1, 2, 0]}

In [44]:
df = pd.DataFrame(data)

In [45]:
df

text  label
0       The quick brown fox jumps over the lazy dog.      0
1  Artificial intelligence is transforming the wo...      1
2        Hugging Face makes working with NLP easier.      2
3  Streaming datasets allow training on large cor...      0
4        Python is my favorite programming language.      1
5     Open source contributions accelerate research.      2
6  Data engineers love clean and structured pipel...      0
7   Transformers are powerful models for many tasks.      1
8  Machine learning requires both theory and prac...      2
9  Randomly generated text can still be useful fo...      0

In [46]:
# Converting the dataframe in a format which is uploadable to huggingface as a dataset.

dataset = Dataset.from_pandas(df)

In [47]:
dataset

Dataset({
    features: ['text', 'label'],
    num_rows: 10
})

We do this step so that the label column is converted to huggingface compatible encoded column.

In [48]:
dataset = dataset.class_encode_column("label")

Stringifying the column:   0%|          | 0/10 [00:00<?, ? examples/s]

Casting to class labels:   0%|          | 0/10 [00:00<?, ? examples/s]

In [49]:
dataset

Dataset({
    features: ['text', 'label'],
    num_rows: 10
})

Always pass the huggingface write token if you want to make any changes to your huggingface account.

In [50]:
from huggingface_hub import notebook_login

notebook_login()

In [51]:
from huggingface_hub import whoami
print(whoami())

{'type': 'user', 'id': '671f58d678721464275180e7', 'name': 'AtharvaBOT7', 'fullname': 'Atharva Chundurwar', 'email': 'atharvachundurwar841@gmail.com', 'emailVerified': True, 'canPay': False, 'periodEnd': None, 'isPro': False, 'avatarUrl': 'https://cdn-avatars.huggingface.co/v1/production/uploads/no-auth/p2Tv-NKUAtssx-fz5Q5th.png', 'orgs': [], 'auth': {'type': 'access_token', 'accessToken': {'displayName': 'testingwrite', 'role': 'write', 'createdAt': '2025-10-01T18:54:53.908Z'}}}


Pushing the custom created dataset to hugging face datasets.

In [52]:
dataset.push_to_hub("AtharvaBOT7/testinghf")

Uploading the dataset shards:   0%|          | 0/1 [00:00<?, ?it/s]

Creating parquet from Arrow format:   0%|          | 0/1 [00:00<?, ?ba/s]

Uploading files as a binary IO buffer is not supported by Xet Storage. Falling back to HTTP upload.


README.md:   0%|          | 0.00/379 [00:00<?, ?B/s]

No files have been modified since last commit. Skipping to prevent empty commit.


CommitInfo(commit_url='https://huggingface.co/datasets/AtharvaBOT7/testinghf/commit/0ef437db1a25e778808c83df61e82f318943dcf5', commit_message='Upload dataset', commit_description='', oid='0ef437db1a25e778808c83df61e82f318943dcf5', pr_url=None, repo_url=RepoUrl('https://huggingface.co/datasets/AtharvaBOT7/testinghf', endpoint='https://huggingface.co', repo_type='dataset', repo_id='AtharvaBOT7/testinghf'), pr_revision=None, pr_num=None)

# Tokenization

Now we will convert our data into tokens and then pass the data to transformer model.

In [53]:
!pip install transformers

We are now importing the bert-base-cased tokenizer model for tokenizing our dataset.

In [54]:
from transformers import AutoTokenizer, AutoModel
tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

In [55]:
def tokenize(example):
  return tokenizer(example, truncation=True, padding="max_length")

In [56]:
dataset["text"][1]

'Artificial intelligence is transforming the world.'

In [57]:
tokenize(dataset["text"][1])

{'input_ids': [101, 2051, 19814, 2916, 4810, 1110, 20892, 1103, 1362, 119, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [58]:
tokens = tokenizer.tokenize("Hello World, I am Atharva")

In [59]:
tokens

['Hello', 'World', ',', 'I', 'am', 'At', '##har', '##va']

In [60]:
inputs = tokenizer("Hello World, I am Atharva", return_tensors="pt")

In [61]:
inputs

{'input_ids': tensor([[ 101, 8667, 1291,  117,  146, 1821, 1335, 7111, 2497,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}

Now we will map the entire data.

In [62]:
def tokenize1(example):
  return tokenizer(example["text"], truncation=True, padding="max_length")

In [63]:
encoded_dataset = dataset.map(tokenize1, batched=True)

Map:   0%|          | 0/10 [00:00<?, ? examples/s]

In [64]:
encoded_dataset

Dataset({
    features: ['text', 'label', 'input_ids', 'token_type_ids', 'attention_mask'],
    num_rows: 10
})

## Trick to tokenize large datasets faster

We can see that when we use the parameter use_fast = True we get around 7 seconds of execution time and when we keep it to False, we get around 17 seconds of execution time.

This is because when we use use_fast = True the backend code is implemented using Rust language and when we use use_fast = False, python code is run in backend.

In [65]:
import time
text = ["This is a sample sentence for tokenization."] * 100000

In [66]:
# Fast method

start = time.time()
tok_fast = AutoTokenizer.from_pretrained("bert-base-cased", use_fast=True)
tok_fast(text, padding=True, truncation=True)
print("Fast time: ", time.time() - start)

Fast time:  8.45846939086914


In [67]:
# Normal / slow method

start = time.time()
tok_slow = AutoTokenizer.from_pretrained("bert-base-cased", use_fast=False)
tok_slow(text, padding=True, truncation=True)
print("Slow time: ", time.time() - start)

Slow time:  18.781182050704956


# Creating Your own Tokenizer from scratch

The most famous algorithm on HuggingFace for creating your own tokenizer is BPE (Byte Pair Encoding) and it is used alongside Hugging Face's fast tokenizer library.

BPE is a subword tokenization algorithm that helps break down words into smaller pieces (subwords) making it useful for handling rarte words and unknown words in NLP tasks.

In [68]:
from tokenizers import Tokenizer, models, pre_tokenizers, trainers
corpus = ["The quick brown fox jumps over the lazy dog.", "Artificial intelligence is transforming the world.", "Hugging Face makes working with NLP easier. \n"]

In [69]:
corpus

['The quick brown fox jumps over the lazy dog.',
 'Artificial intelligence is transforming the world.',
 'Hugging Face makes working with NLP easier. \n']

In [70]:
tokenizer = Tokenizer(models.BPE())

Removing the white spaces between the words.

In [71]:
tokenizer.pre_tokenizer = pre_tokenizers.Whitespace()

In [72]:
trainer = trainers.BpeTrainer(vocab_size=50)

In [73]:
tokenizer.train_from_iterator(corpus, trainer)

In [74]:
tokenizer.save("custom-tokenizer.json")

Now Loading the tokenizer that we created just now.

In [75]:
from transformers import PreTrainedTokenizerFast

hf_tokenizer = PreTrainedTokenizerFast(
    tokenizer_file="/content/custom-tokenizer.json",
    unk_token="<unk>",
    pad_token="<pad>",
    cls_token="<cls>",
    sep_token="<sep>",
    mask_token="<mask>"
)

In [76]:
hf_tokenizer.save_pretrained("my-tokenizer-hf")

('my-tokenizer-hf/tokenizer_config.json',
 'my-tokenizer-hf/special_tokens_map.json',
 'my-tokenizer-hf/tokenizer.json')

# Loading the Model

Using the AutoTokenizer class we can load the specific tokenizer for that particular model. Each and every model has been trained on its own tokenizer.

HuggingFace has many AutoModel classes and there maybe a class for your specific task.

In [77]:
from transformers import AutoTokenizer, AutoModel
tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

In [78]:
inputs = tokenizer("Hello World, I am Atharva, how are you?", return_tensors="pt")

In [79]:
model = AutoModel.from_pretrained("bert-base-cased")

In [80]:
model

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(28996, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSdpaSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False

Now we are sending the encoded values as input to our bert base model. This model does not have any output head therefore we will get output embeddings only.

In [81]:
import torch
with torch.no_grad():
  outputs = model(**inputs)

In [82]:
print(outputs)

BaseModelOutputWithPoolingAndCrossAttentions(last_hidden_state=tensor([[[ 0.3367,  0.3648,  0.3177,  ..., -0.2858,  0.3472, -0.0040],
         [ 0.6072, -0.5583,  0.6113,  ..., -0.4887,  0.4278, -0.1739],
         [ 0.0890, -0.0298,  0.1732,  ..., -0.4665,  0.1360, -0.6062],
         ...,
         [ 0.5991,  0.0240,  0.3782,  ..., -0.3929, -0.6011,  0.3424],
         [ 0.3859,  0.0130,  0.4674,  ..., -0.1409, -0.3569,  0.3854],
         [ 0.5125,  0.3729,  0.5536,  ..., -0.8669,  0.2425, -0.7082]]]), pooler_output=tensor([[-8.0949e-01,  4.5041e-01,  9.9995e-01, -9.9625e-01,  9.7732e-01,
          8.6038e-01,  9.9216e-01, -9.8221e-01, -9.8732e-01, -7.2435e-01,
          9.8128e-01,  9.9904e-01, -9.9602e-01, -9.9993e-01,  7.2807e-01,
         -9.8831e-01,  9.9162e-01, -6.2994e-01, -9.9998e-01, -5.3705e-01,
         -4.8222e-01, -9.9995e-01,  4.0080e-01,  9.2432e-01,  9.8709e-01,
          6.2459e-02,  9.8617e-01,  9.9999e-01,  9.4536e-01,  3.5585e-02,
          3.5087e-01, -9.9438e-01,  

In [83]:
print(outputs.last_hidden_state.shape)

torch.Size([1, 15, 768])


# Performing small Practical

In [84]:
from torch.nn.functional import cosine_similarity

In [85]:
sentence1 = "Hello, how are you?"
sentence2 = "Hi, how are you?"

In [86]:
inputs1 = tokenizer(sentence1, return_tensors="pt")
inputs2 = tokenizer(sentence2, return_tensors="pt")

In [89]:
with torch.no_grad():
  output1 = model(**inputs1).last_hidden_state[0].mean(dim=0)
  output2 = model(**inputs2).last_hidden_state[0].mean(dim=0)

In [91]:
similarity = cosine_similarity(output1,output2,dim=0)
print(f"Similarity: {similarity.item():4f}")

Similarity: 0.987065


# Installing Sentence Transformers

In [92]:
!pip install -U sentence-transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.6/486.6 kB 20.4 MB/s eta 0:00:00
  Attempting uninstall: sentence-transformers
    Found existing installation: sentence-transformers 5.1.0
    Uninstalling sentence-transformers-5.1.0:
      Successfully uninstalled sentence-transformers-5.1.0


In [93]:
from sentence_transformers import SentenceTransformer

# Loading the sentence embedding model

In [94]:
model = SentenceTransformer("all-MiniLM-L6-v2")

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

## Step 1: Encode both the sentences into embeddings

In [95]:
embedding1 = model.encode(sentence1, convert_to_tensor=True)
embedding2 = model.encode(sentence2, convert_to_tensor=True)

In [97]:
len(embedding1)
len(embedding2)

384

In [96]:
similarity = cosine_similarity(embedding1,embedding2,dim=0)
print(f"Similarity: {similarity.item():4f}")

Similarity: 0.947376


Earlier when we performed the small practical we go the similarity to be very high because the cosine_similarity function took all the inputs into account but the all-mini lm model only took 384 features into account.

In [106]:
sentence = "I am very happy"

In [107]:
inputs = tokenizer(sentence, return_tensors="pt")

In [108]:
inputs

{'input_ids': tensor([[ 101,  146, 1821, 1304, 2816,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1]])}

In [109]:
inputs.pop("token_type_ids")

tensor([[0, 0, 0, 0, 0, 0]])

In [110]:
from transformers import AutoModelForSequenceClassification

In [111]:
clf_model = AutoModelForSequenceClassification.from_pretrained("distilbert-base-uncased-finetuned-sst-2-english")

In [112]:
with torch.no_grad():
  outputs = clf_model(**inputs)

In [113]:
logits = outputs.logits

In [114]:
predicted_class = torch.argmax(logits, dim=1)

In [115]:
print(predicted_class.item())

0


In [116]:
from transformers import AutoTokenizer, AutoModelForCausalLM

In [117]:
gpt_tok = AutoTokenizer.from_pretrained("gpt2")
gpt_model = AutoModelForCausalLM.from_pretrained("gpt2")

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

In [118]:
sentence = "Transformers are best"

In [119]:
gpt_inputs = gpt_tok(sentence, return_tensors="pt")

In [120]:
with torch.no_grad():
  gpt_outputs = gpt_model.generate(
      gpt_inputs["input_ids"],
      max_length=gpt_inputs["input_ids"].shape[1] + 5,
      do_sample=False
  )

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


In [121]:
generated_text = gpt_tok.decode(gpt_outputs[0], skip_special_tokens=True)

In [122]:
print("GPT-2 generated: ", generated_text)

GPT-2 generated:  Transformers are best used to create a new


# AutoConfig class loads the configuration of any model

In [123]:
from transformers import AutoConfig, AutoModel

In [124]:
config = AutoConfig.from_pretrained("bert-base-cased")

In [125]:
print(config.hidden_size)
print(config.num_attention_heads)
print(config.num_hidden_layers)
print(config.hidden_act)
print(config.num_labels)
print(config.vocab_size)

768
12
12
gelu
2
28996


Suppose we dont want to use this model for binary classification and use it for multi class classification, then we can simply change the variables.

In [126]:
config.num_labels = 5

In [127]:
model = AutoModel.from_config(config)

In [128]:
model

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(28996, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSdpaSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False

# To download the model from HuggingFace on local

In [129]:
from huggingface_hub import snapshot_download

This will download the model in the local directory.

In [130]:
snapshot_download(
    repo_id="bert-base-uncased",
    local_dir="/content/bert-base-uncased",
    local_dir_use_symlinks=False
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/file_download.py:982: UserWarning: `local_dir_use_symlinks` parameter is deprecated and will be ignored. The process to download files to a local folder has been updated and do not rely on symlinks anymore. You only need to pass a destination folder as`local_dir`.
For more details, check out https://huggingface.co/docs/huggingface_hub/main/en/guides/download#download-files-to-local-folder.
  warnings.warn(


Fetching 16 files:   0%|          | 0/16 [00:00<?, ?it/s]

flax_model.msgpack:   0%|          | 0.00/438M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.mlmodel:   0%|          | 0.00/165k [00:00<?, ?B/s]

Manifest.json:   0%|          | 0.00/617 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

coreml/fill-mask/float32_model.mlpackage(…):   0%|          | 0.00/532M [00:00<?, ?B/s]

LICENSE:   0%|          | 0.00/11.4k [00:00<?, ?B/s]

.gitattributes:   0%|          | 0.00/491 [00:00<?, ?B/s]

model.onnx:   0%|          | 0.00/532M [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/440M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

rust_model.ot:   0%|          | 0.00/534M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

tf_model.h5:   0%|          | 0.00/536M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

'/content/bert-base-uncased'

Whenever we download the model, it's tokenizer will get installed automatically.

# Loading the downloaded model

In [131]:
from transformers import AutoTokenizer, AutoModel

In [132]:
tokenizer = AutoTokenizer.from_pretrained("/content/bert-base-uncased")

In [133]:
model = AutoModel.from_pretrained("/content/bert-base-uncased")

# Using the Inbuilt Pipeline Method

In [134]:
from transformers import pipeline

In [137]:
classifier = pipeline("sentiment-analysis")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

Device set to use cpu


In [139]:
classifier("I want to learn everyhting that the world has to offer")

[{'label': 'POSITIVE', 'score': 0.9926204681396484}]

In [140]:
classifier("I am a bad boy")

[{'label': 'NEGATIVE', 'score': 0.9919100403785706}]

We can assign tokens to the Zero Shot Classifier model

In [144]:
new_classifier = pipeline("zero-shot-classification")

No model was supplied, defaulted to facebook/bart-large-mnli and revision d7645e1 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cpu


In [145]:
res = new_classifier(
    "I am trying to learn about the Transformer models in python.",
    candidate_labels=["education", "politics", "business"]
)

In [146]:
print(res)

{'sequence': 'I am trying to learn about the Transformer models in python.', 'labels': ['education', 'business', 'politics'], 'scores': [0.8161661624908447, 0.11951912194490433, 0.06431470811367035]}


Loading the Generation pipeline

In [148]:
generation = pipeline("text-generation")

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cpu


In [149]:
generation("I am a good boy, what do you think?")

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'I am a good boy, what do you think? Because I don\'t know. I have no idea. But if you would like to help me, you can help me. I\'m not here to help anybody. You are here to help me. And the only way you can help me is if you help me. When someone says, "You\'re right," you have a right to say it. But when someone says, "I\'m right," you have nothing to say. You have nothing to say anything. You can\'t stop me. You can\'t stop anybody. You can\'t stop the poor person, the poor man. You can\'t stop the poor man, the poor man, the poor man. You have nothing to say. You can\'t stop anybody. You have nothing to say anything. You can\'t stop the poor man. You have nothing to say anything. You can\'t stop the poor man. You have nothing to say anything. You can\'t stop the poor man. You have nothing to say anything. You can\'t stop the poor man. You have nothing to say anything. You can\'t stop the poor man. You have nothing to say anything. You can\'t stop the poor man. Y

just like we performed generation task and the zero shot classification task, we can also perform summarization, question-answering and many other important day to day tasks.


# Evaluation Metrics

In [150]:
! pip install evaluate rouge_score evaluate[visualization]

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 6.1 MB/s eta 0:00:00
  Created wheel for rouge_score: filename=rouge_score-0.1.2-py3-none-any.whl size=24934 sha256=0f3a52b93d190dead28babbac720b39b2eab84ae7a0ade29a50176d3933f2207
  Stored in directory: /root/.cache/pip/wheels/85/9d/af/01feefbe7d55ef5468796f0c68225b6788e85d9d0a281e7a70
Successfully built rouge_score


In [151]:
import evaluate

In [152]:
accuracy = evaluate.load("accuracy")

In [153]:
result = accuracy.compute(predictions=[0,1,1,0], references=[0,1,0,0])

In [154]:
print(result)

{'accuracy': 0.75}


In [155]:
def compute_metrics(eval_pred):
  logits, labels = eval_pred
  predictions = logits.argmax(axis=-1)
  return accuracy.compute(predictions=predictions, references=labels)

In [156]:
from datasets import load_dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer

In [157]:
train_dataset = load_dataset("imdb", split="train[:100]")
test_dataset = load_dataset("imdb", split="test[:100]")

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

In [158]:
model = AutoModelForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels = 2)
tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [159]:
def tokenize(example):
  return tokenizer(example["text"], truncation=True, padding="max_length")

In [160]:
train_encoded = train_dataset.map(tokenize, batched=True)
test_encoded = test_dataset.map(tokenize, batched=True)

Map:   0%|          | 0/100 [00:00<?, ? examples/s]

Map:   0%|          | 0/100 [00:00<?, ? examples/s]

This is the encoded format used for Torch.

In [161]:
train_encoded.set_format("torch", columns=["input_ids", "attention_mask", "label"])
test_encoded.set_format("torch", columns=["input_ids", "attention_mask", "label"])

In [163]:
args = TrainingArguments(
    output_dir="eval-check",
    per_device_eval_batch_size=32,
    report_to="none"
)

In [164]:
trainer = Trainer(
    model=model,
    args=args,
    compute_metrics=compute_metrics
)

In [165]:
metrics = trainer.evaluate(eval_dataset=test_encoded)

/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:666: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  warnings.warn(warn_msg)


In [166]:
print(metrics)

{'eval_loss': 0.7059656381607056, 'eval_model_preparation_time': 0.0017, 'eval_accuracy': 0.15, 'eval_runtime': 100.2355, 'eval_samples_per_second': 0.998, 'eval_steps_per_second': 0.04}
